### PR TITLE
Implement ranges::generate and ranges::generate_n

### PR DESCRIPTION
**Pull request by ahanamuk**

### CODE REVIEW

File: stl/inc/algorithm [hunk 1]
```diff
@@ -2348,6 +2348,61 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
...
 
+#ifdef __cpp_lib_concepts
+namespace ranges {
+    // VARIABLE ranges::generate
+    class _Generate_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+            template <input_or_output_iterator _O, sentinel_for<_O> _S, copy_constructible _F>
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:2359]: Should de-indent this code once; occurs below.

Comment (StephanTLavavej) [inline stl/inc/algorithm:2359]: Should rename single-letter identifiers; occurs below.

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 2]
```diff
@@ -0,0 +1,34 @@
+#include <algorithm>
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:1]: Missing copyright/license banner.

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#include <concepts>
+#include <range_algorithm_support.hpp>
+#include <ranges>
+#include <utility>
+
+constexpr void smoke_test() {
+    using ranges::generate_n, ranges::iterator_t;
+    using std::same_as;
+
+    int output1[] = {13, 42, 1367};
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:12]: Should this be `output` as there is only one?

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+#include <utility>
+
+constexpr void smoke_test() {
+    using ranges::generate, ranges::iterator_t;
+    using std::same_as;
+
+    int output1[] = {13, 42, 1367};
+    int output2[] = {13, 42, 1367};
+    {
+        auto result = generate(std::begin(output1), std::end(output1), []() { return 7; });
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:18]: Pure style, so your call: the `()` is optional in lambda expressions like these that take no arguments and don't need to specify a return type, or `mutable`, or any of the other bits of syntax that go before or after the `()`. You could write simply `[] { return 7; }` here if you like.

Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:18]: Should we use something more complicated than a constant sequences to verify that the algorithm makes no extraneous calls to the generator function?

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+        for (int i = 0; i < 3; ++i) {
```
Comment (CaseyCarter) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:19]: As in the `fill` review, these could be range-based for loops.

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp (hunk 3 above)
Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:12]: Ditto "should we unbrace the codeblock"?

**Comment by CaseyCarter** [comment]
Details on the Code Format Validation test says there's trailing whitespace somewhere in `<algorithm>`.

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp (hunk 4 above)
Comment (mnatsuhara) [reply tests/std/tests/P0896R4_ranges_alg_generate/test.cpp]: @ahanamuk I think the simplest way to verify no extraneous calls are being made to the generator function would be to have the generator function use a static variable.  For example, the function could increment the variable before returning it in each call so the value it returns is directly tied to how many times the function has been called.  (e.g., instead of the lambda returning `3` every time it is called, it would first return `1` when called the first time, then `2` when called the second time, then `3` on the third and if it is somehow being called more than `N` times, you would be able to tell because the returned value would be `4` or greater.)  I found [this](https://stackoverflow.com/questions/8391058/how-do-static-variables-in-lambda-function-objects-work) explanation to be helpful in terms of a statics + lambdas explanation.  
@CaseyCarter Please correct me if I'm advising poorly/misinterpreting your suggestion!

File: stl/inc/algorithm [hunk 6]
```diff
@@ -2348,6 +2348,61 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
...
+
+    // VARIABLE ranges::generate_n
+    class _Generate_n_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <input_or_output_iterator _Out, copy_constructible _Fn>
+            requires invocable<_Fn&> && indirectly_writable<_Out, invoke_result_t<_Fn&>>
+        constexpr _Out operator()(_Out _First, iter_difference_t<_Out> _Count, _Fn _Gen) const {
```
Comment (mnatsuhara) [inline stl/inc/algorithm:2391]: Similar to Casey's comment [here](https://github.com/microsoft/STL/pull/904#discussion_r441920864) should we avoid all of this work if `_Count <= 0`?  Similar to the non-ranges version of the algorithm:
```
// replace [_Dest, _Dest + _Count) with _Func()
_Algorithm_int_t<_Diff> _Count = _Count_raw;
if (0 < _Count) {
   ...
}

return _Dest;
```
Though I believe this is a "yoda condition" (a term I recently learned) that I don't know that we need to replicate :P

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+#include <range_algorithm_support.hpp>
+#include <ranges>
+#include <utility>
+
+constexpr void smoke_test() {
+    using ranges::generate, ranges::iterator_t;
+    using std::same_as;
+
+    int output1[] = {13, 42, 1367};
+    int output2[] = {13, 42, 1367};
```
Comment (mnatsuhara) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:16]: Again, Casey [commented](https://github.com/microsoft/STL/pull/904#discussion_r441925510) similarly on the `fill`/`fill_n` PR but since `output1` is only used in the first code block and `output2` only used in the second, can these be moved into more local scope?

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        assert(result == std::end(output));
+    }
+    {
+        int output[]          = {13, 42, 1367};
+        int calls_to_generate = -1;
+        auto result           = generate(output, [&calls_to_generate]() {
+            calls_to_generate++;
+            return calls_to_generate;
+        });
+        for (int i = 0; i < 3; i++) {
```
Comment (mnatsuhara) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:40]:
```suggestion
        for (int i = 0; i < 3; ++i) {
```
Just to beat STL to the comment 😄

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+        for (auto elem : output) {
+            assert(elem == value);
+        }
+        assert(result == std::end(output));
+    }
+    {
+        int output[]          = {13, 42, 1367};
+        int calls_to_generate = -1;
+        auto result           = generate(output, [&calls_to_generate]() {
+            calls_to_generate++;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:38]:
```suggestion
            ++calls_to_generate;
```

File: stl/inc/algorithm [hunk 10]
```diff
@@ -2841,6 +2841,61 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
...
+            }
+
+            _Seek_wrapped(_First, _STD move(_UFirst));
+            return _First;
+        }
+
+        template <class _Rng, copy_constructible _Fn>
+            requires invocable<_Fn&> && output_range<_Rng, invoke_result_t<_Fn&>>
+        constexpr borrowed_iterator_t<_Rng> operator()(_Rng&& _Range, _Fn _Gen) const {
+            return (*this)(_RANGES begin(_Range), _RANGES end(_Range), _Gen);
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:2869]: I believe this should use `_Pass_fn`, which is an internal helper to reduce the expense of passing function objects within STL machinery. (In the classic C++98/03 era, function objects were assumed to be cheap to copy, but that is less true these days given the runtime expense of `std::function` and similar machinery.) I observe that every other `_Rng&& _Range` overload uses `_Pass_fn` except for `for_each` which is special (it uses `_STD move` because its function object must ultimately be returned).

```suggestion
            return (*this)(_RANGES begin(_Range), _RANGES end(_Range), _Pass_fn(_Gen));
```

File: stl/inc/algorithm (hunk 6 above)
Comment (StephanTLavavej) [reply stl/inc/algorithm:2884]: Excellent find - yes, I believe this is needed for correctness, not just performance.

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+
+constexpr void smoke_test() {
+    using ranges::generate, ranges::iterator_t;
+    using std::same_as;
+
+    {
+        int output[]    = {13, 42, 1367};
+        const int value = 7;
+        auto result     = generate(std::begin(output), std::end(output), []() { return value; });
+        for (auto elem : output) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:20]: As mentioned in https://github.com/microsoft/STL/pull/904#discussion_r443809863 :

```suggestion
        for (const auto& elem : output) {
```

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+            assert(elem == value);
+        }
+        assert(result == std::end(output));
+    }
+    {
+        int output[]    = {13, 42, 1367};
+        const int value = 13;
+        auto result     = generate(output, []() { return value; });
+        for (auto elem : output) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:29]:
```suggestion
        for (const auto& elem : output) {
```

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+#include <range_algorithm_support.hpp>
+
+constexpr void smoke_test() {
+    using ranges::generate_n, ranges::iterator_t;
+    using std::same_as;
+
+    int output[]    = {13, 42, 1367};
+    const int value = 7;
+    auto result     = generate_n(std::begin(output), ranges::distance(output), []() { return value; });
+    for (auto elem : output) {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:19]:
```suggestion
    for (const auto& elem : output) {
```

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+        assert(elem == value);
+    }
+    assert(result == std::end(output));
+}
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:23]: Should we also test `generate_n` with the "incrementing lambda" technique?

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+            assert(elem == value);
+        }
+        assert(result == std::end(output));
+    }
+    {
+        int output[]          = {13, 42, 1367};
+        int calls_to_generate = -1;
+        auto result           = generate(output, [&calls_to_generate]() {
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:37]: This is great! It made me realize that there's one more way that the implementation could be incorrect, and how to verify it cheaply. It's tempting for us to mark things as `const`, but we really don't want to mark the generator function object as `const` in the algorithm (which would prohibit function objects with non-`const` function call operators).

Here, the lambda function has a `const` function call operator (like all lambdas receive by default); it modifies `calls_to_generate` through the stored reference. That's fine, except that it wouldn't detect the potential implementation problem of marking the function object as `const`. Since we don't need `calls_to_generate` outside the lambda, we could instead use an init-capture and make the lambda `mutable` (which removes `const` from its function call operator). The behavior would be the same, but this would verify that `generate()` is invoking the function object without adding `const` to it. I believe the syntax for this is:

```suggestion
        auto result           = generate(output, [calls_to_generate = -1]() mutable {
```

(This will also affect clang-format.)

**Comment by StephanTLavavej** [comment]
Oops, got confused about whether I had previously approved this.

File: stl/inc/algorithm [hunk 16]
```diff
@@ -3003,6 +3003,63 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
...
+    // VARIABLE ranges::generate_n
+    class _Generate_n_fn : private _Not_quite_object {
+    public:
+        using _Not_quite_object::_Not_quite_object;
+
+        // clang-format off
+        template <input_or_output_iterator _Out, copy_constructible _Fn>
+            requires invocable<_Fn&> && indirectly_writable<_Out, invoke_result_t<_Fn&>>
+        constexpr _Out operator()(_Out _First, iter_difference_t<_Out> _Count, _Fn _Gen) const {
+            if (0 <= _Count) {
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:3047]: I believe this should be strictly greater-than (since there's nothing to do for generating N=0 elements), and non-Yoda:

```suggestion
            if (_Count > 0) {
```

File: stl/inc/algorithm [hunk 17]
```diff
@@ -3003,6 +3003,63 @@ _FwdIt2 replace_copy_if(_ExPo&&, _FwdIt1 _First, _FwdIt1 _Last, _FwdIt2 _Dest, _
...
+            }
+
+            _Seek_wrapped(_First, _STD move(_UFirst));
+            return _First;
+        }
+
+        template <class _Rng, copy_constructible _Fn>
+            requires invocable<_Fn&> && output_range<_Rng, invoke_result_t<_Fn&>>
+        constexpr borrowed_iterator_t<_Rng> operator()(_Rng&& _Range, _Fn _Gen) const {
+            return (*this)(_RANGES begin(_Range), _RANGES end(_Range), _Pass_fn(_Gen));
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:3031]: Should this eventually follow a more complicated pattern (which I believe I've seen in other ranges PRs) to avoid unnecessarily passing wrapped iterators to `operator()`? No change requested for this PR if this is a todo recorded somewhere.

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 18]
```diff
@@ -0,0 +1,60 @@
...
+#include <cassert>
+#include <concepts>
+#include <ranges>
+#include <utility>
+
+#include <range_algorithm_support.hpp>
+
+constexpr void smoke_test() {
+    using ranges::generate, ranges::iterator_t;
+    using std::same_as;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:14]: Unused using-declarations:

```suggestion
    using ranges::generate;
```

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+#include <cassert>
+#include <concepts>
+#include <ranges>
+#include <utility>
+
+#include <range_algorithm_support.hpp>
+
+constexpr void smoke_test() {
+    using ranges::generate_n, ranges::iterator_t;
+    using std::same_as;
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:14]:
```suggestion
    using ranges::generate_n;
```

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 20]
```diff
@@ -0,0 +1,39 @@
...
+    int output[] = {13, 42, 1367};
+    auto result  = generate_n(ranges::begin(output), ranges::distance(output), [calls_to_generate = -1]() mutable {
+        ++calls_to_generate;
+        return calls_to_generate;
+    });
+    for (int i = 0; i < 3; ++i) {
+        assert(i == output[i]);
+    }
+    assert(result == ranges::end(output));
+}
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:25]: Should we also have test cases for zero and negative N?

File: stl/inc/algorithm (hunk 17 above)
Comment (CaseyCarter) [reply stl/inc/algorithm:3031]: > No change requested for this PR if this is a todo recorded somewhere.

I have a TODO to refactor all Ranges algorithms for perf and codesize; I'm happy enough for now that this is simple and correct.

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 21]
```diff
@@ -0,0 +1,59 @@
...
+        assert(result == ranges::end(output));
+    }
+    {
+        int expected_output[] = {13, 42, 1367};
+        int output[]          = {13, 42, 1367};
+        auto result           = generate_n(ranges::begin(output), 0, [calls_to_generate = -1]() mutable {
+            ++calls_to_generate;
+            return calls_to_generate;
+        });
+        assert(ranges::equal(output, expected_output));
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:32]: Line 13 is `using` `ranges::equal`:
```suggestion
        assert(equal(output, expected_output));
```

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 22]
```diff
@@ -0,0 +1,59 @@
...
+        assert(result == ranges::begin(output));
+    }
+    {
+        int expected_output[] = {13, 42, 1367};
+        int output[]          = {13, 42, 1367};
+        auto result           = generate_n(ranges::begin(output), -1, [calls_to_generate = -1]() mutable {
+            ++calls_to_generate;
+            return calls_to_generate;
+        });
+        assert(ranges::equal(output, expected_output));
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:42]:
```suggestion
        assert(equal(output, expected_output));
```

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+        int output[]          = {13, 42, 1367};
+        auto result           = generate_n(ranges::begin(output), 0, [calls_to_generate = -1]() mutable {
+            ++calls_to_generate;
+            return calls_to_generate;
+        });
+        assert(ranges::equal(output, expected_output));
+        assert(result == ranges::begin(output));
+    }
+    {
+        int expected_output[] = {13, 42, 1367};
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:36]: Ditto, `expected_output` should be `const`, `constexpr`, or `static constexpr`.

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+            ++calls_to_generate;
+            return calls_to_generate;
+        });
+        for (int i = 0; i < 3; ++i) {
+            assert(i == output[i]);
+        }
+        assert(result == ranges::end(output));
+    }
+    {
+        int expected_output[] = {13, 42, 1367};
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:26]: This `expected_output` should be `const`, `constexpr`, or `static constexpr`, which helps to prevent unintentionally modifying the expected results.

(I had typed up a long explanation of the differences, but apparently I made a mistake by not actually saving that as a comment, and then GitHub allowed me to submit my earlier review by dropping that pending comment entirely, or I am confused. Anyways, the differences are minor for test code; `static constexpr` is the most efficient and can be used here.)

File: stl/inc/algorithm [hunk 25]
```diff
@@ -3163,12 +3163,64 @@ namespace ranges {
 
                 _Seek_wrapped(_First, _STD move(_UFirst));
             }
-
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/algorithm]: Nope.

File: tests/std/include/range_algorithm_support.hpp [hunk 26]
```diff
@@ -807,7 +807,7 @@ struct with_input_ranges {
     }
 };
 
-template <class Continuation, class Element>
+template <class Continuation, class Element = int>
```
[the remark this answers is not in the document]
Comment (CaseyCarter) [reply tests/std/include/range_algorithm_support.hpp]: I can now officially confirm there are too many Ranges PRs. I convinced myself while resolving merge conflicts that this PR was removing template default arguments having forgotten that we already merged #959.

[the remark this answers is not in the document]
Comment (miscco) [reply tests/std/include/range_algorithm_support.hpp]: sorry? 🙃

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply tests/std/include/range_algorithm_support.hpp]: > sorry? 🙃

No need to be. I just need to do a better job of streamlining the review and merge process so we don't have a dozen PRs open at a time ;)

File: stl/inc/algorithm [hunk 27]
```diff
@@ -3169,6 +3169,74 @@ namespace ranges {
...
+        // clang-format off
+        template <input_or_output_iterator _Out, copy_constructible _Fn>
+            requires invocable<_Fn&> && indirectly_writable<_Out, invoke_result_t<_Fn&>>
+        constexpr _Out operator()(_Out _First, iter_difference_t<_Out> _Count, _Fn _Gen) const {
+            if (_Count > 0) {
+                auto _UFirst = _Get_unwrapped_n(_STD move(_First), _Count);
+                do {
+                    *_UFirst = _Gen();
+                    ++_UFirst;
+                } while (--_Count > 0);
```
Comment (StephanTLavavej) [inline stl/inc/algorithm:3230]: Nice, this avoids an unnecessary branch. Thanks!

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 28]
```diff
@@ -0,0 +1,53 @@
...
+    static constexpr void call() {
+        using ranges::generate_n, ranges::equal, ranges::iterator_t;
+
+        const auto iota_gen = [count = 0]() mutable { return count++; };
+
+        {
+            int output[] = {13, 42, 1367};
+            Out out_wrapper{output};
+            auto result = generate_n(out_wrapper.begin(), ranges::distance(output), iota_gen);
+            STATIC_ASSERT(same_as<decltype(result), iterator_t<Out>>);
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:24]: I observe that the `generate` test isn't inspecting the iterator type returned by the range overload, like this `generate_n` test is doing. Not sure if they should be consistent.

Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:24]: > Not sure if they should be consistent.

It certainly can't hurt.

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 29]
```diff
@@ -0,0 +1,46 @@
...
+#include <range_algorithm_support.hpp>
+
+using namespace std;
+
+struct instantiator {
+    template <ranges::output_range<const int&> Out>
+    static constexpr void call() {
+        using ranges::generate, ranges::iterator_t;
+
+        const auto iota_gen = [count = 0]() mutable { return count++; };
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:18]: Now that this test contains a `using`-directive, I feel compelled to comment that this init-capture shadows the name of an algorithm.
```suggestion
        const auto iota_gen = [val = 0]() mutable { return val++; };
```

File: tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+#include <range_algorithm_support.hpp>
+
+using namespace std;
+
+struct instantiator {
+    template <ranges::output_range<const int&> Out>
+    static constexpr void call() {
+        using ranges::generate_n, ranges::equal, ranges::iterator_t;
+
+        const auto iota_gen = [count = 0]() mutable { return count++; };
```
Comment (StephanTLavavej) [inline tests/std/tests/P0896R4_ranges_alg_generate_n/test.cpp:18]:
```suggestion
        const auto iota_gen = [val = 0]() mutable { return val++; };
```

File: tests/std/tests/P0896R4_ranges_alg_generate/test.cpp [hunk 31]
```diff
@@ -0,0 +1,46 @@
...
+
+#include <algorithm>
+#include <cassert>
+#include <concepts>
+#include <ranges>
+
+#include <range_algorithm_support.hpp>
+
+using namespace std;
+
```
Comment (mnatsuhara) [inline tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:12]: Last minute note: Should these tests also "Validate dangling story" as many [others ](https://github.com/microsoft/STL/blob/master/tests/std/tests/P0896R4_ranges_alg_is_sorted/test.cpp#L17-L19) do?  Or is there reason these algorithms should be exempt?

Comment (CaseyCarter) [reply tests/std/tests/P0896R4_ranges_alg_generate/test.cpp:12]: Yes! `generate_n` doesn't accept `range` arguments, but `generate` does so it needs some `dangling` coverage.

**Comment by StephanTLavavej** [comment]
Thanks for implementing these ranges algorithms, they're going to *generate* a lot of excitement. :sunglasses:

**Comment by miscco** [comment]
@CaseyCarter how did you hack into @StephanTLavavej's account?

**Comment by CaseyCarter** [comment]
> @CaseyCarter how did you hack into @StephanTLavavej's account?

I had the exact same thought - he's even using the sunglasses emoji instead of cat emoji! I feel seen.
